pkg/tests/api/alerting/api_testing_test.go

```diff
@@ -0,0 +1,399 @@
+		require.Equal(t, http.StatusOK, status)
+		var result []amv2.PostableAlert
+		require.NoErrorf(t, json.Unmarshal([]byte(body), &result), "cannot parse response to data frame")
+		require.Len(t, result, 4)
+		labels := []string{"GA", "FL", "AL", "AZ"}
+		for i, alert := range result {
+			require.Equal(t, labels[i], alert.Labels["state"])
+		}
+	})
+
+	t.Run("valid rule should return built-in labels", func(t *testing.T) {
+		rule := genRule(testDataRule())
+		status, body := apiCli.SubmitRuleForTesting(t, rule)
+		require.Equal(t, http.StatusOK, status)
+		var result []amv2.PostableAlert
+		require.NoErrorf(t, json.Unmarshal([]byte(body), &result), "cannot parse response to data frame")
+		require.Len(t, result, 4)
+		for _, alert := range result {
+			require.Equal(t, rule.Rule.GrafanaManagedAlert.Title, alert.Labels[model.AlertNameLabel])
+			require.Equal(t, rule.NamespaceUID, alert.Labels[alertingModels.NamespaceUIDLabel])
+			require.Equal(t, rule.NamespaceTitle, alert.Labels[ngmodels.FolderTitleLabel])
+		}
+	})
+
+	t.Run("invalid rule should reject request", func(t *testing.T) {
+		req := genRule(alertRuleGen())
+		req.Rule = apimodels.PostableExtendedRuleNode{}
+		status, _ := apiCli.SubmitRuleForTesting(t, req)
+		require.Equal(t, http.StatusBadRequest, status)
+	})
+
+	t.Run("if user does not have permissions", func(t *testing.T) {
+		if !setting.IsEnterprise {
+			t.Skip("Enterprise-only test")
+		}
+
+		testUserId := createUser(t, env.SQLStore, user.CreateUserCommand{
+			DefaultOrgRole: "",
+			Password:       "test",
+			Login:          "test",
+		})
+
+		testUserApiCli := newAlertingApiClient(grafanaListedAddr, "test", "test")
+
+		t.Run("fail if can't read rules", func(t *testing.T) {
+			status, body := testUserApiCli.SubmitRuleForTesting(t, genRule(alertRuleGen()))
+			require.Contains(t, body, accesscontrol.ActionAlertingRuleRead)
+			require.Equalf(t, http.StatusForbidden, status, "Response: %s", body)
+		})
+
+		// access control permissions store
+		permissionsStore := resourcepermissions.NewStore(env.SQLStore)
+		_, err := permissionsStore.SetUserResourcePermission(context.Background(),
+			accesscontrol.GlobalOrgID,
+			accesscontrol.User{ID: testUserId},
+			resourcepermissions.SetResourcePermissionCommand{
+				Actions: []string{
+					accesscontrol.ActionAlertingRuleRead,
+				},
+				Resource:          "folders",
+				ResourceID:        "*",
+				ResourceAttribute: "uid",
+			}, nil)
+		require.NoError(t, err)
+		testUserApiCli.ReloadCachedPermissions(t)
+
+		t.Run("fail if can't query data sources", func(t *testing.T) {
+			status, body := testUserApiCli.SubmitRuleForTesting(t, genRule(alertRuleGen()))
+			require.Contains(t, body, "user is not authorized to query one or many data sources used by the rule")
+			require.Equalf(t, http.StatusUnauthorized, status, "Response: %s", body)
+		})
+	})
+}
+
+func testDataRule() func() apimodels.PostableExtendedRuleNode {
```

Comment:
It looks like this model is used only in the test above, and subtests make assumptions about the content of the rule sent to API. I wonder if it would be cleaner to have this function in the test context only and not share it with other tests.

Reply:
Agreed, I'm iterating on this section right now. Will try to make the function more useful to future test writers.